### ISSUE TITLE
`taskStart` date is not recorded correctly

### ISSUE THREAD

**Issue by manaswis** [ProjectSidewalk/SidewalkWebpage#463]
While working on other issues, I found that the `taskStart` date stored doesn't seem to reflect the actual time the task started in the database. See screenshot below:

<img width="989" alt="screen shot 2016-12-02 at 11 34 22 am" src="https://cloud.githubusercontent.com/assets/2873216/20841965/c1ca113e-b883-11e6-8c13-37185f2954dc.png">

All the tasks have the same `taskStart` date. To the best of my knowledge, we currently aren't using it anywhere (others may comment on this) but this should be fixed if we want to do any task completion based analytics.

**Comment by jonfroehlich** [comment]
And we absolutely will. This seems like a big problem. How can we calculate average task length (in time) without this functioning correctly.

@kotarohara, had you seen this before?

**Comment by manaswis** [comment]
After investigating the issue, I found that what its recording is the "session start date" i.e. the time the `/audit` page was loaded the first time - so it preserves a session. Soheil is currently doing the analysis using the taskStart date stored in the `audit_task_interaction` table. Also, the code isn't straightforward to modify to update the taskStart date.

After contemplating the pros and cons, we feel that at this point, we can keep it as-is. The most we could do is rename the `taskStart` column in the `audit_task` table as `audit_session_start`. That will preserve the consistency in the database as well as make the analysis code simpler - as we won't have to modify the code based on the version of the application. We feel its not worth the time as this point to make changes to incorporate the taskStart logic in the existing framework.

Thoughts?

**Comment by jonfroehlich** [comment]
Sounds good. Should we move this to v3 or just mark this idiosyncrasy is some documentation?

**Comment by manaswis** [comment]
I think adding it to the documentation is best for now; we note there that when we refactor the codebase we should modify the logic then.

**Comment by jonfroehlich** [comment]
OK, in the documentation, link back to this Issue on github.

**Comment by manaswis** [comment]
I have added a note about this issue in the [developer notes](https://github.com/ProjectSidewalk/SidewalkWebpage/wiki/Developer-notes) stored in the wiki of this github repo.

**Comment by manaswis** [comment]
Closing this issue now.

**Comment by misaugstad** [comment]
I just noticed this myself. I'm going to reopen for now, hopefully we can fix it before launching to new cities.

**Comment by misaugstad** [comment]
There is actually a much bigger problem here, where the `task_start` seems to be sometimes created as the time when the tool was first launched (11/17/2015). This is a must-fix before launching to new cities.

![bad_timestamps](https://user-images.githubusercontent.com/6518824/43170760-9cc19d7a-8f5b-11e8-8447-4761fd0f1348.png)

The above comes from various user ids. The query I ran is the following:
```
SELECT task_start, task_end
FROM audit_task
WHERE completed = TRUE
ORDER BY task_start, task_end DESC
```

This has happened to approximately *half* of the audit task entries in the database.

All is not lost, however! The task_end seems to be recorded correctly everywhere, including for the incomplete audit tasks. So we do know roughly when each audit task occurred, we just don't have the exactly start times.

**Comment by ShivenBhatt** [comment]
Ok, took a look at this issue. Here's what I found:
We have two different methods for determining time. We use one for determining the start time, and one for determining the end time.
For the start time, we use:

For the end time, we use:

The method using Calendar isn't thread safe, but the method using DateTime is. Still working on how that difference could cause this issue, or if there are any other differences between Calendar and DateTime.

here's where I found that Calendar isn't thread safe: https://stackoverflow.com/questions/12131324/is-java-util-calendar-thread-safe-or-not

here's where I found that DateTIme is thread safe: http://joda-time.sourceforge.net/apidocs/org/joda/time/DateTime.html

**Comment by misaugstad** [comment]
This is very very promising!! Nice sleuthing @ShivenBhatt :sunglasses: 

It is not immediately obvious how this causes the behavior we have been seeing, but do let us know if you find something that would explain it (or if you find a way to replicate the bug)!

**Comment by jonfroehlich** [comment]
> here's where I found that DateTIme is thread safe: http://joda-time.sourceforge.net/apidocs/org/joda/time/DateTime.html

Wait, are we using the joda-time DateTime object `org.joda.time.DateTime` in Project Sidewalk? If not, then the documentation that @ShivenBhatt linked to is not relevant.

I also don't get the sense that this has anything to do with thread safety given how different the start times and end times are. Why would thread unsafe code end up with a 2015 start date?

**Comment by ShivenBhatt** [comment]
this was in the imports section
```
import org.joda.time.{DateTime, DateTimeZone}
```
so yeah, we're using ```org.joda.time.DateTime```

And the thread safety stuff was just a possible difference that I've found. I'm pretty sure that there has to be something wrong with Calendar, because there isn't any way to modify the task_start column through the AuditTaskTable. Therefore, the time we're passing in when we make the task must be incorrect. Therefore, Calendar must be breaking at some point. Therefore, there has to be a breaking difference between
```
val timestamp: Timestamp = new Timestamp(Calendar.getInstance(TimeZone.getTimeZone("UTC")).getTime.getTime)
```
and
```
val now = new DateTime(DateTimeZone.UTC)
val timestamp: Timestamp = new Timestamp(now.getMillis)
```
In other words, I've found the problem line of code, I just don't know why or when it's breaking. I'm going to try to repro these issues with calendar by writing some java tests.

Also, we really shouldn't be using calendar in this way anyway. Currently, we're getting a Date object from the calendar (```Calendar.getInstance(TimeZone.getTimeZone("UTC")).getTime```)and then getting the time in milliseconds from the Date object by calling ```getTime``` again. Date has been depreciated since 1997. We should be doing ```Calendar.getInstance(TimeZone.getTimeZone("UTC")).getTimeInMillis```
Also, we shouldn't even be using Calendar. Since Java 8, we have a new Date API that is thread safe and reportedly more accurate.

**Comment by jonfroehlich** [comment]
Great work @ShivenBhatt. Are we using Java 8? If so, I agree that we should move to upgrade all of our time usage stuff to using the new Date API and state-of-the-art date handling conventions.

**Comment by ShivenBhatt** [comment]
@jonfroehlich, you're right, the problem isn't that Calendar isn't thread safe. Looks like there's more going on here. I ran this basic java code separately to test how the Calendar was working with threads:
```
package testcals;
import java.sql.*;
import java.util.*;
public class TestCals {
    public static boolean stopSpamThreads = false;
    public static SpamTimestampThread[] threads = new SpamTimestampThread[100];
    public static void main(String[] args) {
       
       // TODO code application logic here
       for(int i = 0; i < threads.length; i++){
           //create the thread, run the thread
           threads[i] = new SpamTimestampThread();
           threads[i].threadname = i + "";
           threads[i].start();
       }    
       //make scanner to detect console input
       Scanner in = new Scanner(System.in);
       //hang here until we enter in a newline
       in.nextLine();
       //once the user makes a newline, stop all the threads.
       stopSpamThreads = true;
    }
    
    public static class SpamTimestampThread extends Thread {
        String threadname = "";
        int iteration = 0;
        @Override
        public void run(){
           while(!stopSpamThreads){
               //make timestamp from calendar 
               Timestamp timestamp = new Timestamp(Calendar.getInstance(TimeZone.getTimeZone("UTC")).getTime().getTime());
               System.out.println("Thread name:\t" + threadname + ",\ttime:\t" + timestamp + ",\titeration:\t" + iteration);
               iteration++;
           }
        }
    }    
}
``` 
and I go this data:
```
Thread name:	95,	time:	2018-08-07 10:12:58.493,	iteration:	143
Thread name:	95,	time:	2018-08-07 10:12:58.538,	iteration:	144
Thread name:	90,	time:	2018-08-07 10:12:58.493,	iteration:	120
Thread name:	46,	time:	2018-08-07 10:12:58.493,	iteration:	184
Thread name:	34,	time:	2018-08-07 10:12:58.493,	iteration:	115
Thread name:	34,	time:	2018-08-07 10:12:58.538,	iteration:	116
Thread name:	34,	time:	2018-08-07 10:12:58.538,	iteration:	117
Thread name:	44,	time:	2018-08-07 10:12:58.493,	iteration:	129
Thread name:	2,	time:	2018-08-07 10:12:58.493,	iteration:	158
Thread name:	16,	time:	2018-08-07 10:12:58.493,	iteration:	130
Thread name:	59,	time:	2018-08-07 10:12:58.493,	iteration:	160
Thread name:	59,	time:	2018-08-07 10:12:58.538,	iteration:	161
Thread name:	64,	time:	2018-08-07 10:12:58.493,	iteration:	265
Thread name:	23,	time:	2018-08-07 10:12:58.493,	iteration:	226
Thread name:	39,	time:	2018-08-07 10:12:58.493,	iteration:	167
Thread name:	82,	time:	2018-08-07 10:12:58.493,	iteration:	133
Thread name:	4,	time:	2018-08-07 10:12:58.493,	iteration:	175
Thread name:	5,	time:	2018-08-07 10:12:58.493,	iteration:	192
Thread name:	56,	time:	2018-08-07 10:12:58.493,	iteration:	154
Thread name:	92,	time:	2018-08-07 10:12:58.493,	iteration:	139
Thread name:	63,	time:	2018-08-07 10:12:58.493,	iteration:	152
Thread name:	18,	time:	2018-08-07 10:12:58.493,	iteration:	182
Thread name:	65,	time:	2018-08-07 10:12:58.493,	iteration:	216
Thread name:	87,	time:	2018-08-07 10:12:58.493,	iteration:	139
Thread name:	13,	time:	2018-08-07 10:12:58.493,	iteration:	125
Thread name:	49,	time:	2018-08-07 10:12:58.493,	iteration:	123
Thread name:	98,	time:	2018-08-07 10:12:58.493,	iteration:	128
Thread name:	14,	time:	2018-08-07 10:12:58.493,	iteration:	124
Thread name:	9,	time:	2018-08-07 10:12:58.493,	iteration:	112
Thread name:	64,	time:	2018-08-07 10:12:58.538,	iteration:	266
Thread name:	59,	time:	2018-08-07 10:12:58.538,	iteration:	162
Thread name:	16,	time:	2018-08-07 10:12:58.538,	iteration:	131
Thread name:	2,	time:	2018-08-07 10:12:58.538,	iteration:	159
Thread name:	44,	time:	2018-08-07 10:12:58.538,	iteration:	130
Thread name:	34,	time:	2018-08-07 10:12:58.538,	iteration:	118
Thread name:	46,	time:	2018-08-07 10:12:58.538,	iteration:	185
Thread name:	90,	time:	2018-08-07 10:12:58.538,	iteration:	121
Thread name:	95,	time:	2018-08-07 10:12:58.538,	iteration:	145
Thread name:	79,	time:	2018-08-07 10:12:58.538,	iteration:	117
Thread name:	86,	time:	2018-08-07 10:12:58.538,	iteration:	142
Thread name:	77,	time:	2018-08-07 10:12:58.537,	iteration:	122
Thread name:	15,	time:	2018-08-07 10:12:58.537,	iteration:	141
Thread name:	84,	time:	2018-08-07 10:12:58.537,	iteration:	209
Thread name:	33,	time:	2018-08-07 10:12:58.537,	iteration:	146
Thread name:	69,	time:	2018-08-07 10:12:58.537,	iteration:	155
Thread name:	54,	time:	2018-08-07 10:12:58.537,	iteration:	114
Thread name:	3,	time:	2018-08-07 10:12:58.537,	iteration:	198
Thread name:	74,	time:	2018-08-07 10:12:58.537,	iteration:	178
Thread name:	12,	time:	2018-08-07 10:12:58.537,	iteration:	127
Thread name:	43,	time:	2018-08-07 10:12:58.537,	iteration:	193
Thread name:	17,	time:	2018-08-07 10:12:58.537,	iteration:	138
Thread name:	32,	time:	2018-08-07 10:12:58.537,	iteration:	143
Thread name:	51,	time:	2018-08-07 10:12:58.537,	iteration:	180
Thread name:	94,	time:	2018-08-07 10:12:58.537,	iteration:	137
Thread name:	27,	time:	2018-08-07 10:12:58.537,	iteration:	121
Thread name:	99,	time:	2018-08-07 10:12:58.537,	iteration:	136
Thread name:	52,	time:	2018-08-07 10:12:58.537,	iteration:	128
Thread name:	24,	time:	2018-08-07 10:12:58.537,	iteration:	148
Thread name:	42,	time:	2018-08-07 10:12:58.537,	iteration:	130
Thread name:	25,	time:	2018-08-07 10:12:58.537,	iteration:	134
Thread name:	66,	time:	2018-08-07 10:12:58.537,	iteration:	222
Thread name:	93,	time:	2018-08-07 10:12:58.537,	iteration:	147
Thread name:	73,	time:	2018-08-07 10:12:58.537,	iteration:	131
Thread name:	30,	time:	2018-08-07 10:12:58.537,	iteration:	163
Thread name:	36,	time:	2018-08-07 10:12:58.537,	iteration:	144
Thread name:	22,	time:	2018-08-07 10:12:58.537,	iteration:	136
Thread name:	37,	time:	2018-08-07 10:12:58.537,	iteration:	146
Thread name:	45,	time:	2018-08-07 10:12:58.537,	iteration:	130
Thread name:	76,	time:	2018-08-07 10:12:58.537,	iteration:	158
Thread name:	8,	time:	2018-08-07 10:12:58.537,	iteration:	121
Thread name:	75,	time:	2018-08-07 10:12:58.537,	iteration:	140
Thread name:	21,	time:	2018-08-07 10:12:58.537,	iteration:	143
Thread name:	62,	time:	2018-08-07 10:12:58.537,	iteration:	148
Thread name:	71,	time:	2018-08-07 10:12:58.537,	iteration:	164
Thread name:	60,	time:	2018-08-07 10:12:58.536,	iteration:	154
Thread name:	68,	time:	2018-08-07 10:12:58.536,	iteration:	136
Thread name:	80,	time:	2018-08-07 10:12:58.523,	iteration:	141
Thread name:	11,	time:	2018-08-07 10:12:58.523,	iteration:	147
Thread name:	72,	time:	2018-08-07 10:12:58.523,	iteration:	212
Thread name:	6,	time:	2018-08-07 10:12:58.523,	iteration:	189
Thread name:	38,	time:	2018-08-07 10:12:58.523,	iteration:	130
Thread name:	85,	time:	2018-08-07 10:12:58.523,	iteration:	129
Thread name:	26,	time:	2018-08-07 10:12:58.523,	iteration:	149
Thread name:	53,	time:	2018-08-07 10:12:58.523,	iteration:	184
Thread name:	88,	time:	2018-08-07 10:12:58.523,	iteration:	286
Thread name:	81,	time:	2018-08-07 10:12:58.523,	iteration:	151
Thread name:	89,	time:	2018-08-07 10:12:58.522,	iteration:	159
Thread name:	10,	time:	2018-08-07 10:12:58.522,	iteration:	167
Thread name:	31,	time:	2018-08-07 10:12:58.522,	iteration:	133
Thread name:	58,	time:	2018-08-07 10:12:58.522,	iteration:	169
Thread name:	97,	time:	2018-08-07 10:12:58.522,	iteration:	118
Thread name:	55,	time:	2018-08-07 10:12:58.522,	iteration:	161
Thread name:	7,	time:	2018-08-07 10:12:58.522,	iteration:	171
Thread name:	61,	time:	2018-08-07 10:12:58.522,	iteration:	150
Thread name:	47,	time:	2018-08-07 10:12:58.522,	iteration:	155
Thread name:	91,	time:	2018-08-07 10:12:58.522,	iteration:	151
Thread name:	35,	time:	2018-08-07 10:12:58.522,	iteration:	118
Thread name:	28,	time:	2018-08-07 10:12:58.522,	iteration:	151
Thread name:	48,	time:	2018-08-07 10:12:58.522,	iteration:	190
Thread name:	1,	time:	2018-08-07 10:12:58.522,	iteration:	134
Thread name:	57,	time:	2018-08-07 10:12:58.522,	iteration:	138
Thread name:	29,	time:	2018-08-07 10:12:58.522,	iteration:	111
Thread name:	50,	time:	2018-08-07 10:12:58.522,	iteration:	120
Thread name:	41,	time:	2018-08-07 10:12:58.522,	iteration:	114
Thread name:	40,	time:	2018-08-07 10:12:58.522,	iteration:	133
Thread name:	83,	time:	2018-08-07 10:12:58.522,	iteration:	154
Thread name:	67,	time:	2018-08-07 10:12:58.509,	iteration:	134
Thread name:	70,	time:	2018-08-07 10:12:58.509,	iteration:	125
Thread name:	78,	time:	2018-08-07 10:12:58.509,	iteration:	130
Thread name:	20,	time:	2018-08-07 10:12:58.509,	iteration:	125
Thread name:	19,	time:	2018-08-07 10:12:58.509,	iteration:	133
Thread name:	96,	time:	2018-08-07 10:12:58.509,	iteration:	107
Thread name:	0,	time:	2018-08-07 10:12:58.509,	iteration:	156
```
The iterations are very different from each other, but the time seems to be accurate down to the second (after the tenth of a second, it begins to vary).

This effect, however, doesn't explain the problem. So it's back to square one.

**Comment by ShivenBhatt** [comment]
If that's the case, we should just copy over this method of finding the time:
```
val now = new DateTime(DateTimeZone.UTC)
val timestamp: Timestamp = new Timestamp(now.getMillis)
```
since it seems to work for the end time stamps.

The reason I was trying to test the Calendar this way was because I wasn't able to replicate the issue on my end. Since I can't replicate it, I can't test if this fix works any better than Calendar. I thought that if I could prove Calendar was causing the issue, I could give a better rationale for the change.

**Comment by misaugstad** [comment]
>The reason I was trying to test the Calendar this way was because I wasn't able to replicate the issue on my end. Since I can't replicate it, I can't test if this fix works any better than Calendar. I thought that if I could prove Calendar was causing the issue, I could give a better rationale for the change.

Yep we do still need to either have the ability to replicate the bug (and thus prove that it has been fixed) _or_ find an adequate explanation for why this change would fix the bug, you're right about that Shiven :)

And nice work so far! I'm really into the testing! :+1: :+1:

**Comment by ShivenBhatt** [comment]
AHA!

The error seems to build up over time. I let the test sit for 30 minutes on accident :) and the error went from a fraction of a second to three minutes...

```
Thread name:	69304,	time:	2018-08-07 13:54:52.745,	iteration:	3
Thread name:	69304,	time:	2018-08-07 13:57:19.045,	iteration:	4
Thread name:	69304,	time:	2018-08-07 13:57:19.046,	iteration:	5
Thread name:	69304,	time:	2018-08-07 13:57:19.046,	iteration:	6
Thread name:	69304,	time:	2018-08-07 13:57:19.046,	iteration:	7
Thread name:	69299,	time:	2018-08-07 13:54:52.745,	iteration:	14
Thread name:	69302,	time:	2018-08-07 13:54:52.745,	iteration:	1
Thread name:	69302,	time:	2018-08-07 13:57:19.047,	iteration:	2
Thread name:	69302,	time:	2018-08-07 13:57:19.047,	iteration:	3
Thread name:	69302,	time:	2018-08-07 13:57:19.047,	iteration:	4
Thread name:	69302,	time:	2018-08-07 13:57:19.047,	iteration:	5
Thread name:	69301,	time:	2018-08-07 13:54:52.745,	iteration:	1
Thread name:	69298,	time:	2018-08-07 13:54:52.745,	iteration:	3
Thread name:	69305,	time:	2018-08-07 13:54:52.745,	iteration:	1
Thread name:	69295,	time:	2018-08-07 13:54:52.745,	iteration:	6
Thread name:	69078,	time:	2018-08-07 13:54:52.745,	iteration:	1
```
On top of that, we're starting to get the freezing behavior that we saw in the first post. Note The last 5 entries. 

I think this is reason enough to get rid of calendar, and replace it with the thread-safe, working call that we use for end time stamps.

**Comment by misaugstad** [comment]
This is a great find! And glad to see you created a new issue specific to this! Because I'm still not convinced that this is the reason for the bug that this thread is about, but it definitely needs to be addressed either way!

**Comment by misaugstad** [comment]
While working on #324 I stumbled upon what's going wrong here. On the front end, whenever a new task/street is started, we are never recording the time when the task starts (not sure why this wasn't the first thing that was checked).

When the Explore page loads, we send Task object for every street in the neighborhood. Most of the fields for the Task are correct, like the street geometry or street audit priority. But some of them should be filled in or updated as users actually audit that street. The `task_start` value and `task_end` values are the key ones that should be updated.

We never update the `task_start`. We are just filling in a dummy timestamp there on the back end with the current time at the beginning of the session, and then the front end just sends that number right back to the back end when users actually start auditing the street.

Currently making the fix in #2979
